#!/bin/bash
# Before running crossmap, we should get bedgraph files containing k-mer mappabilities with k-mer length 75 for exons and 36 for UTRs. These results can be generated by genmap.
mkdir -p crossmap log
jsub -q fat -n 2 -R "span[hosts=1]" -J sheep.crossmap.1 -e log/crossmap.1.%J.log -o log/crossmap.1.%J.log "bash crossmap.1.sh"

# actually compute cross-mappability (-initonly FALSE)
comp_dir="/storage/public/home/2020060185/genome/sheep/reference/mappability/crossmap"
script_dir="$comp_dir/script"
mappability_dir="$comp_dir/gene_mappability"
mappability_fn="$mappability_dir/gene_mappability.txt"
cross_mappability_dir="$comp_dir/cross_mappability"
n_gene_per_crossmap_batch=2000
n_gene_in_mappability_file=$(wc -l $mappability_fn | sed 's/ .*//g')
for n1 in $(seq 1 $n_gene_per_crossmap_batch $n_gene_in_mappability_file)
do
  n2=$(($n1+$n_gene_per_crossmap_batch-1))
  script_fn="${script_dir}/compute_cross_mappability_2_${n1}_${n2}.sh"
  jsub -q normal -n 2 -R "span[hosts=1]" -J sheep.crossmap.2_${n1}_${n2} -e log/crossmap.2.${n1}_${n2}.%J.log -o log/crossmap.2.${n1}_${n2}.%J.log "bash $script_fn"
done

# combine all cross-mappability results into one file
combined_cross_mappability_fn="$comp_dir/cross_mappability.txt"
if [ -f $combined_cross_mappability_fn ]; then rm "$combined_cross_mappability_fn";  fi
for fn in "$cross_mappability_dir"/*/*.crossmap.txt
do
  cat $fn >> "$combined_cross_mappability_fn"
done
